### PR TITLE
Business Rules Manager for WSO2 Stream Processor

### PR DESCRIPTION
**Pull request by minudika**
## Purpose
In Data Analytics, there are common use cases for analyzing statistics that involve operations such as calculating the average, minimum, maximum etc., for different endpoints. Business Rules should allow user to use templates in order to generate business rules needed for different scenarios with common requirements in a convenient manner.
Domain-specific parameter values (e.g., attribute name, time duration) should be able to be customized and business 
rules should be able to be generated.
This PR is for adding feature for managing business rules in WSO2 Stream Processor.
## Goals
This feature allows,
 * Developers to add and configure templates in file system level.
 * Business Users to create a business rule from an existing template or from scratch and later, list,edit and delete existing business rules through UI.
## Approach
The developer will be organizing template groups and putting into the resources directory. The business user then is 
allowed to create a Business Rule from template, or a Business Rule from scratch. The UI Provides the ability to 
choose a template group, and select one from the available rule templates under the selected template group. Then 
he/she is redirected to a form, where values are filled and SiddhiApps are generated from them. These saved Business 
Rules can be accessed later for purposes of viewing, editing or deleting. 
## User stories
* As a developer I want to add a template into the stream processor and configure it , so that a user can create business rules from it.
* As a Business User I should be able to create a Business Rule from an existing template
  so that I can use it to process events
* As a Business User I should be able to view Business Rules that are configured So that I can manage a Business Rule from it
* As a Business User I should be able to modify a business rule so that the existing business rule can perform with 
updated values
* As a Business User I should be able to delete a business rule so that I can remove unwanted business rules
* As a Business User I should be able to create a Business Rule from the scratch, so that I can use it to process events

Find complete user stories  [here](https://docs.google.com/document/d/1w1w6VDagWLaOpkCzsz5adjLx8UhTeHkzS641dSM3pjQ/edit?usp=sharing)

### CODE REVIEW

File: components/org.wso2.carbon.business.rules.core/pom.xml [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+        <export.package>
+            org.wso2.carbon.business.rules.core.*
+        </export.package>
+        <import.package>
+            org.osgi.framework.*;version="${osgi.framework.import.version.range}",
+            org.wso2.carbon.kernel;version="${carbon.kernel.package.import.version.range}",
+            org.wso2.msf4j.*;version="${msf4j.import.version.range}",
+            javax.management.*,
+            javax.ws.rs.*;version="0.0.0",
+            org.wso2.siddhi.*;version="4.0.0",
```
Comment (suhothayan) [inline components/org.wso2.carbon.business.rules.core/pom.xml]: any reason to import a specific version ?

File: components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/bean/BusinessRule.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+/**
+ * Represents the parent of Business Rules
+ * created from Template
+ * and created from scratch
+ */
+public abstract class BusinessRule {
+    private String uuid;
+    private String name;
+    private String templateGroupUUID;
+    private String type; // "template" or "scratch"
```
Comment (suhothayan) [inline components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/bean/BusinessRule.java]: cant we make this an enum?

File: components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/bean/RuleTemplate.java [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+import java.util.Collection;
+import java.util.Map;
+
+/**
+ * Represents a Rule Template, which consists of one or more Templates
+ */
+public class RuleTemplate {
+    private String uuid;
+    private String name;
+    private String type; // "input" "output" or "template"
```
Comment (suhothayan) [inline components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/bean/RuleTemplate.java]: cant we make this enum?

File: components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/bean/RuleTemplate.java [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+    private String instanceCount; // "one" or "many"
```
Comment (suhothayan) [inline components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/bean/RuleTemplate.java]: cant we make this enum?

File: components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/datasource/QueryExecutor.java [hunk 5]
```diff
@@ -0,0 +1,391 @@
...
+                    String inputRuleTemplateUUID = jsonObject.get("inputRuleTemplateUUID").getAsString();
+                    String outputRuleTemplateUUID = jsonObject.get("outputRuleTemplateUUID").getAsString();
+                    BusinessRuleFromScratchProperty properties = new Gson().fromJson(jsonObject.get("properties"),
+                            BusinessRuleFromScratchProperty.class);
+                    BusinessRule businessRule = new BusinessRuleFromScratch(uuid, name, templateGroupUUID, type,
+                            inputRuleTemplateUUID, outputRuleTemplateUUID, properties);
+                    Object[] objects = new Object[2];
+                    objects[0] = businessRule;
+                    objects[1] = deploymentStatus;
+                    list.add(objects);
```
Comment (suhothayan) [inline components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/datasource/QueryExecutor.java]: use a pojo intered of using arrays to pass data

File: components/org.wso2.carbon.business.rules.web/README.md [hunk 6]
```diff
@@ -0,0 +1,2154 @@
+This project was bootstrapped with [Create React App](https://github.com/facebookincubator/create-react-app).
+
```
Comment (suhothayan) [inline components/org.wso2.carbon.business.rules.web/README.md]: Why we have such as big readme file ?

File: components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/bean/RuleTemplate.java (hunk 3 above)
Comment (AnuGayan) [reply components/org.wso2.carbon.business.rules.core/src/main/java/org/wso2/carbon/business/rules/core/bean/RuleTemplate.java]: We only use this class to map the ruleTemplate  json, that is reading from the business rule json file. We never manually assign any value to 'type'. So do we need a enum for this?

File: components/org.wso2.carbon.business.rules.web/README.md (hunk 6 above)
Comment (minudika) [reply components/org.wso2.carbon.business.rules.web/README.md]: Updated and added only necessary content to this file.

**Comment by minudika** [comment]
@suhothayan 
Have done required changes. Closing this PR and will re-open with related resources.

**Pull request closed without merging**